prime-router/src/main/kotlin/cli/tests/BasicTests.kt

```diff
@@ -66,172 +66,94 @@ class Ping : CoolTest() {
-            for (result in routeResults.values)
-                passed = passed && examineStepResponse(result, "route", sender.topic)
-            if (!passed) {
-                bad("***async end2end_up FAILED***: Route result invalid")
-            }
-            val receiverNames = routeResults.values.flatMap { it?.reports ?: emptyList() }
-                .map { "${it.receivingOrg}.${it.receivingOrgSvc}" }
-            expectedReceivers.forEach { receiver ->
-                if (!receiverNames.contains(receiver.fullName)) {
-                    bad("***async end2end_up FAILED***: expected ${receiver.fullName} to have received a report")
+     override val name = "end2end_up"
+     override val description = "e2e tests that verify reports are correctly received and uploaded"
+     override val status = TestStatus.SMOKE
+
+     override suspend fun run(environment: Environment, options: CoolTestOptions): Boolean {
+         initListOfGoodReceiversAndCountiesForUniversalPipeline()
+
+         ugly("Running $name asynchronously")
+         var passed = true
+
+         ugly("Starting $name Test: send ${fullELRSender.fullName} data to $allGoodCounties")
+         passed = passed and end2EndUP(
+             environment,
+             options,
+             fullELRSender,
+             listOf(hl7FullELRReceiver, fhirFullELRReceiver)
+         )
+
+         ugly("Starting $name Test: send ${elrElimsSender.fullName} data to $allGoodCounties")
+         passed = passed and end2EndUP(environment, options, elrElimsSender, listOf(elimsReceiver))
+
+         return passed
+     }
+
+     private suspend fun end2EndUP(
+         environment: Environment,
+         options: CoolTestOptions,
+         sender: Sender,
+         expectedReceivers: List<Receiver>,
+     ): Boolean {
+         var passed = true
+         val file = File("src/test/resources/fhirengine/smoketest/valid_hl7.hl7")
+         echo("Loaded datafile $file")
+
+         val (responseCode, json) =
+             HttpUtilities.postReportFile(
+                 environment,
+                 file,
+                 sender,
+                 true,
+                 options.key,
+                 payloadName = "$name ${status.description}",
+             )
+         if (responseCode != HttpURLConnection.HTTP_CREATED) {
+             bad("***end2end_up Test FAILED***: Response code $responseCode")
+             passed = false
+         } else {
+             good("Posting of async report succeeded with response code $responseCode")
+         }
+
+         echo(json)
+         passed = passed and examinePostResponse(json, false)
+         if (!passed) {
+             bad("***end2end_up FAILED***: Error in post response")
+         }
+
+         val reportId = getReportIdFromResponse(json)
+             ?: return bad("***end2end_up FAILED***: Did not find report id")
+
+         // Need just a little wait for the db to catch up
+         waitABit(10, environment)
```

Comment:
Based on my understanding of how [waitABit](https://github.com/CDCgov/prime-reportstream/blob/791e8b0c79c97238239ea894bda6541c590be3b8/prime-router/src/main/kotlin/cli/tests/TestReportStream.kt#L449) works, this should be fine. **However**, there may be edge cases in how the UP system works that this won't be a long enough wait for, or it may be way too long of a wait. Opinions welcome.

Reply:
So as alternative you could use the pattern [here](https://github.com/CDCgov/prime-reportstream/blob/1e725139b15d2438bd2e743bf6f377fa5a58baf0/prime-router/src/main/kotlin/cli/tests/SyncTranslationSchemasTest.kt#L116) which is essentially just a coroutine with a timeout.

It might be a little more reliable since it can complete quicker if the report is ready.

I also don't think the comments in `waitABit` actually work since the output is printed after the tests are done so you don't actually get feedback